extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -663,34 +660,35 @@ public String getEarliestSequenceNumber(StreamPartition<String> partition)
     return getSequenceNumber(partition, ShardIteratorType.TRIM_HORIZON);
   }
 
+  /**
+   * Use the API listShards which is the recommended way instead of describeStream
+   * listShards can return 1000 shards per call and has a limit of 100TPS
+   * This makes the method resilient to LimitExceeded exceptions (compared to 100 shards, 10 TPS of describeStream)
+   *
+   * @param stream name of stream
+   *
+   * @return Set of Shard ids
+   */
   @Override
   public Set<String> getPartitionIds(String stream)
   {
-    return wrapExceptions(
-        () -> {
-          final Set<String> retVal = new HashSet<>();
-          DescribeStreamRequest request = new DescribeStreamRequest();
-          request.setStreamName(stream);
-
-          while (request != null) {
-            final DescribeStreamResult result = kinesis.describeStream(request);
-            final StreamDescription streamDescription = result.getStreamDescription();
-            final List<Shard> shards = streamDescription.getShards();
-
-            for (Shard shard : shards) {
-              retVal.add(shard.getShardId());
-            }
-
-            if (streamDescription.isHasMoreShards()) {
-              request.setExclusiveStartShardId(Iterables.getLast(shards).getShardId());
-            } else {
-              request = null;
-            }
-          }
-
+    return wrapExceptions(() -> {
+      final Set<String> retVal = new HashSet<>();
```

Comment:
can make this object immutable via an `ImmutableSet`

Reply:
this is getting modified later in the code. do you mean returning an immutable copy?

Reply:
yes, I meant that so that outside this method the object is not mutated

Reply:
ah ok. this is a local variable and this class is not keeping any reference to this set or is using it for anything else. so making it immutable doesn't seem necessary.

Reply:
Yes, the variable is local but the set does get passed to the caller via a public method. The idea was that unless we're ok with the callers mutating the set, we can pass them an immutable set so that no caller mutates the set.
Yes, certainly it's not necessary - the initial comment was intended more towards practice and not anything critical